### PR TITLE
[db io managers] support static partitions

### PR DESCRIPTION
**Pull request by jamiedemaria**
### Summary & Motivation
Adds static partition support for DBIOManager base class. Updates duckdb and snowflake io managers accordingly

docs plan - i'm going to add multi partition support, then do a pass at updating api docs and the guides to cover partitions

### How I Tested These Changes
bk - new unit tests

### CODE REVIEW

**Comment by jamiedemaria** [comment]
Current dependencies on/for this PR:
* master
  * **PR #12129** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/12129" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/12129?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/storage/db_io_manager.py
Comment (sryza) [inline python_modules/dagster/dagster/_core/storage/db_io_manager.py:31]: I would lean towards combining these into a single `partition` field that's `Union[TimeWindow, str]`, because exactly one of them must be set.

Comment (jamiedemaria) [reply python_modules/dagster/dagster/_core/storage/db_io_manager.py:31]: ok - i think it would still need a boolean or something to indicate if it's time or static partitioning so that we can make the correct select statement (== vs between). but then both of those would be required, which would be good

Comment (jamiedemaria) [reply python_modules/dagster/dagster/_core/storage/db_io_manager.py:31]: isinstance(str) would also be fine

**Comment by sryza** [comment]
@jamiedemaria want to unmark as a draft and request a review from me when you want me to take another look?

File: python_modules/libraries/dagster-duckdb/dagster_duckdb/io_manager.py
Comment (pzarabadip) [inline python_modules/libraries/dagster-duckdb/dagster_duckdb/io_manager.py:152]: Hi @jamiedemaria 
As I mentioned in Slack about the performance, when I was testing my implementation of static and multi-dimensional partitions, I notice a huge performance drop (from 5 seconds materialisation to 5 minutes and more). I tracked it down to this line and using `DELETE FROM` statement which apparently becomes significant when we deal with large number of rows. In our case, this is the case where we deal with static partitions. Cutting long story short, I could regain the performance by bypassing the direct `DELETE` through creating an new table with rows which we want to keep, dropping the original table, and then recreating the original table from the temporary table, and finally dropping the temporary table. My `_get_cleanup_statement` and `_get_where_clause` functions look like below:
```
def _get_cleanup_statement(table_slice: TableSlice) -> str:
    """
    Returns a SQL statement that deletes data in the given table to make way for the output data
    being written.
    """
    if table_slice.partition:
        where = _get_where_clause(table_slice.partition, delete=True)
        st0 = """DROP TABLE IF EXISTS temporary;"""
        st1 = f"""CREATE TABLE temporary as select * FROM {table_slice.schema}.{table_slice.table} {where};\n"""
        st2 = f"""DROP TABLE {table_slice.schema}.{table_slice.table};\n"""
        st3 = f"""CREATE TABLE {table_slice.schema}.{table_slice.table} AS SELECT * FROM temporary;\n"""
        st4 = """DROP TABLE temporary;"""
        return st0, st1, st2, st3, st4
        # return (
        #     f"DELETE FROM {table_slice.schema}.{table_slice.table}\n"
        #     + _get_where_clause(table_slice.partition)
        # )
    else:
        return f"DELETE FROM {table_slice.schema}.{table_slice.table}"

def _get_where_clause(table_partition: TablePartition, delete: bool = False) -> str:
    time_partitioned: bool = False
    static_partitioned: bool = False
    multi_partitioned: bool = False
    start_dt_str: str = None
    end_dt_str: str = None
    static_part_str: str = None
    DUCKDB_DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S"
    log = get_dagster_logger()
    log.info(table_partition)

    if table_partition.time_window:
        start_dt, end_dt = table_partition.time_window
        start_dt_str = start_dt.strftime(DUCKDB_DATETIME_FORMAT)
        end_dt_str = end_dt.strftime(DUCKDB_DATETIME_FORMAT)
        time_window_key = table_partition.time_window_key
        time_partitioned = True

    if table_partition.static_part:
        static_part_str = table_partition.static_part_key
        static_part_curr_key = int(table_partition.static_part_curr_key)
        static_partitioned = True
    
    if static_partitioned and time_partitioned:
        multi_partitioned = True

    if time_partitioned:
        if delete:
            start = f"""WHERE {time_window_key} <= '{start_dt_str}' """
            end = f"""OR {time_window_key} > '{end_dt_str}'"""
            where_clause = start + end
        else:
            start = f"""WHERE {time_window_key} >= '{start_dt_str}' """
            end = f"""AND {time_window_key} < '{end_dt_str}'"""
            where_clause = start + end
    elif static_partitioned:
        if not delete:
            where_clause = f"""WHERE {static_part_str} != {static_part_curr_key}"""
        else:
            where_clause = f"""WHERE {static_part_str} = {static_part_curr_key}"""
    elif multi_partitioned:
        if delete:
            start_col = f"""WHERE {static_part_str} != {static_part_curr_key} """
            start = f"""({time_window_key} <= '{start_dt_str}' """
            end = f"""OR {time_window_key} > '{end_dt_str}') """
            where_clause = start_col + start + end
        else:
            start_col = f"""WHERE {static_part_str} = {static_part_curr_key} """
            start = f"""{time_window_key} >= '{start_dt_str}' """
            end = f"""AND {time_window_key} < '{end_dt_str}' """
            where_clause = start_col + start + end
    else:
        raise ValueError("I just ran out of options!")
    log.info(where_clause)
    return where_clause
```
I currently have a `delete` argument for testing purposes but this can be improved, if these make sense :)

Comment (jamiedemaria) [reply python_modules/libraries/dagster-duckdb/dagster_duckdb/io_manager.py:152]: thank you for the detail - this is really good to know! I think for the sake of keeping this PR easy to review, i'll keep the delete statement as is and then look to incorporate this in a follow up PR

**Comment by pzarabadip** [comment]
Thanks @jamiedemaria  and @sryza. Sort of related note to this PR (possibly for future PR), about supporting multi-dimensional partitions, I currently have the following function in `db_io_manager.py`,
```
def _get_partition_defs(self, context: OutputContext):
        multi_part: bool = False
        time_window: Optional[TimeWindow] = None
        time_window_key: str = None
        static_part: Optional[StaticPartitionsDefinition] = None
        static_part_key: str = None
        static_part_curr_key: str = None
        output: dict = {}

        log = get_dagster_logger()
        
        if context.has_asset_partitions:
            if isinstance(context.asset_partitions_def, MultiPartitionsDefinition):
                multi_part = True
                part_defs = context.asset_partitions_def
            elif isinstance(context.asset_partitions_def, TimeWindowPartitionsDefinition):
                time_window = context.asset_partitions_time_window
                # log.info(context.asset_partition_key_range)
                time_window_key = context.metadata.get("partition_expr")
            elif isinstance(context.asset_partitions_def, StaticPartitionsDefinition):
                static_part_key = context.metadata.get("partition_expr")
                static_part = context.asset_partitions_def
                static_part_curr_key = context.asset_partition_key
            else:
                raise DagsterInvalidDefinitionError(
                    f"You provided a partition definion ({type(context.asset_partitions_def)}) "
                    f"which is not support by this IOManager"
                )
        
        if multi_part:
            for part in part_defs.partitions_defs:
                if isinstance(part.partitions_def, TimeWindowPartitionsDefinition):
                    time_window = self._get_time_window(context, part.partitions_def)
                    time_window_key = part.name
                if isinstance(part.partitions_def, StaticPartitionsDefinition):
                    static_part = part.partitions_def
                    static_part_key = part.name
                    static_part_curr_key = self._get_static_part_key(context)

        output = {
            'time_window': time_window,
            'time_window_key': time_window_key,
            'static_part': static_part,
            'static_part_key': static_part_key,
            'static_part_curr_key': static_part_curr_key,
            'partitioned': context.has_asset_partitions,
            # 'multi_partitioned': multi_part,
            # 'time_partitioned': time_part,
            # 'static_partitioned': static
        }
        log.info(f'Full output is {output}')
        return output
```
which I call it within `_get_table_slice`, where I can construct my `TablePartition` as follows:
```
part_defs = self._get_partition_defs(output_context)
if part_defs['partitioned']:
            partition = TablePartition(
                time_window=part_defs['time_window'],
                time_window_key=part_defs['time_window_key'],
                static_part=part_defs['static_part'],
                static_part_key=part_defs['static_part_key'],
                static_part_curr_key=part_defs['static_part_curr_key']
            )
```
This implementation is currently working for our cases, although I can make it cleaner. I thought to share it here for further discussion and possibly opening a follow up PR once this one is merged? (I'd be happy to initiate one, if this approach makes sense and is inline with `Dagster` design philosophy). Cheers.

**Review by sryza — APPROVED** [review]
Awesome

**Pull request merged**
